File: Java/src/test/java/uk/org/winder/maths/Fibonacci_TestNG.java
package uk.org.winder.maths;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.function.LongFunction;

import org.testng.annotations.Test;
import org.testng.annotations.DataProvider;

import static org.testng.Assert.assertEquals;

public class Fibonacci_TestNG {
    private Object[][] positiveData = new Object[][]{
            {0L, BigInteger.valueOf(0)},
            {1L, BigInteger.valueOf(1)},
            {2L, BigInteger.valueOf(1)},
            {3L, BigInteger.valueOf(2)},
            {4L, BigInteger.valueOf(3)},
            {5L, BigInteger.valueOf(5)},
            {6L, BigInteger.valueOf(8)},
            {7L, BigInteger.valueOf(13)},
            {8L, BigInteger.valueOf(21)},
            {9L, BigInteger.valueOf(34)},
            {10L, BigInteger.valueOf(55)},
            {11L, BigInteger.valueOf(89)},
            {12L, BigInteger.valueOf(144)},
            {13L, BigInteger.valueOf(233)},
            {14L, BigInteger.valueOf(377)},
            {15L, BigInteger.valueOf(610)},
            {16L, BigInteger.valueOf(987)},
            {17L, BigInteger.valueOf(1597)},
            {18L, BigInteger.valueOf(2584)},
            {19L, BigInteger.valueOf(4181)},
            {20L, BigInteger.valueOf(6765)},
    };

    private Object[] negativeData = new Object[]{-1L, -2L, -5L, -10L, -20L, -100L};

    private Object[] algorithms = new Object[]{
            (LongFunction<BigInteger>) Fibonacci::iterative,
            (LongFunction<BigInteger>) Fibonacci::naiveRecursive,
            (LongFunction<BigInteger>) Fibonacci::memoizedRecursive,
            // (LongFunction<BigInteger>) Fibonacci::closedForm,  // There seems to be a problem with the closedForm implementation.
    };

    @DataProvider
    private Object[][] algorithmsAndPositiveData() {
        final var values = new ArrayList<Object[]>();
        for (var f: algorithms) {
            for (var items: positiveData) {
                if (items.length != 2) {
                    throw new RuntimeException("positiveData array borked.");
                }
                values.add(new Object[]{f, items[0], items[1]});
            }
        }
        return values.toArray(new Object[3][0]);
    }

    @DataProvider
    private Object[][] algorithmsAndNegativeData() {
        final var values = new ArrayList<Object[]>();
        for (var f: algorithms) {
            for (var item: negativeData) {
                values.add(new Object[]{f, item});
            }
        }
        return values.toArray(new Object[3][0]);
    }

    @Test(dataProvider = "algorithmsAndPositiveData")
    public void positiveArgumentsShouldWork(LongFunction<BigInteger> f, long n, BigInteger expected) {
        assertEquals(f.apply(n), expected);
    }

    @Test(dataProvider = "algorithmsAndNegativeData", expectedExceptions = {IllegalArgumentException.class})
    public void negativeArgumentsShouldThrowException(LongFunction<BigInteger> f, long n) {
        f.apply(n);
    }

    @Test
    public void javaStaticAccess() {
        assertEquals(Fibonacci.Sequence.getAt(7), BigInteger.valueOf(13));
        assertEquals(Fibonacci.Sequence.getAt(4), BigInteger.valueOf(3));
        assertEquals(Fibonacci.Sequence.getAt(7), BigInteger.valueOf(13));
        assertEquals(Fibonacci.Sequence.getAt(12), BigInteger.valueOf(144));
        assertEquals(Fibonacci.Sequence.getAt(8), BigInteger.valueOf(21));
    }

    @Test
    public void javaInstanceAccess() {
        final Fibonacci.Sequence fs = new Fibonacci.Sequence();
        assertEquals(fs.getAt(7), BigInteger.valueOf(13));
        assertEquals(fs.getAt(4), BigInteger.valueOf(3));
        assertEquals(fs.getAt(7), BigInteger.valueOf(13));
        assertEquals(fs.getAt(12), BigInteger.valueOf(144));
        assertEquals(fs.getAt(8), BigInteger.valueOf(21));
    }

    @Test
    public void javaIteratorFunction() {
        final var fs = Fibonacci.Sequence.iterator();
        for (var i = 0; i < positiveData.length; ++i) {
            assertEquals(fs.next(), positiveData[i][1]);
        }
    }

    @Test
    public void javaIteratorClassInstance() {
        final var fs = new Fibonacci.Sequence.Iterator();
        for (var i = 0; i < positiveData.length; ++i) {
            assertEquals(fs.next(), positiveData[i][1]);
        }
    }
}

File: Kotlin/src/test/kotlin/uk/org/winder/maths/fibonacci/fibonacci_testng.kt
package uk.org.winder.maths.fibonacci

import java.math.BigInteger

import org.testng.annotations.Test
import org.testng.annotations.DataProvider

import org.testng.Assert.assertEquals

class Fibonacci_TestNG {

  val algorithms = arrayOf(
      "iterative" to {x:Long -> iterative(x)},
      "naïveRecursive" to {x:Long -> naïveRecursive(x)},
      "tailRecursive" to {x:Long -> tailRecursive(x)},
      "sequence" to ::sequence,
      "foldive" to {x:Long -> foldive(x)}
        )

  val positiveData = arrayOf(
      0 to BigInteger.ZERO,
      1 to BigInteger.ONE,
      2 to BigInteger.ONE,
      3 to 2.bigint,
      4 to 3.bigint,
      5 to 5.bigint,
      6 to 8.bigint,
      7 to 13.bigint,
      8 to 21.bigint,
      9 to 34.bigint,
      10 to 55.bigint,
      11 to 89.bigint,
      12 to 144.bigint,
      13 to 233.bigint,
      14 to 377.bigint,
      15 to 610.bigint,
      16 to 987.bigint,
      17 to 1597.bigint,
      18 to 2584.bigint,
      19 to 4181.bigint,
      20 to 6765.bigint
  )

  val negativeData = arrayOf(-1, -2, -5, -10, -20, -100)

  @DataProvider
  fun algorithmsAndPositiveData() = algorithms.flatMap({a -> positiveData.map({d -> arrayOf(a.first, a.second, d.first, d.second)})}).toTypedArray()

  @Test(dataProvider="algorithmsAndPositiveData")
  fun nonNegativeArgument(name:String, algorithm:(Long)->BigInteger, value:Long, expected:BigInteger) {
    assertEquals(algorithm(value), expected, "$name($value)")
  }

  @DataProvider
  fun algorithmsAndNegativeData() = algorithms.flatMap({a -> negativeData.map({d -> arrayOf(a.first, a.second, d)})}).toTypedArray()

  @Test(dataProvider="algorithmsAndNegativeData", expectedExceptions=arrayOf(IllegalArgumentException::class))
  fun negativeArgument(name:String, algorithm:(Long)->BigInteger, value:Long) {
    algorithm(value)
  }

}
